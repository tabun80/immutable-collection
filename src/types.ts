export type CollectionItemType = {
  [key: string]: any;
};

type ArrayType<T extends unknown[]> = T extends (infer U)[] ? U : never;

export type Immutable<T> = T extends unknown[]
  ? readonly Immutable<ArrayType<T>>[]
  : { readonly [K in keyof T]: Immutable<T[K]> };
