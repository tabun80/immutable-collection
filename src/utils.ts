export const deepFreeze = <T extends {} | any[]>(obj: T) => {
  if (isArray(obj)) {
    obj.forEach(deepFreeze);
    return Object.freeze(obj);
  }

  Object.keys(obj).forEach((prop) => {
    const key = prop as keyof T;
    if (typeof obj[key] === 'object' && !Object.isFrozen(obj[key])) {
      deepFreeze(obj[key]);
    }
  });

  return Object.freeze(obj);
};

const isArray = (obj: unknown): obj is any[] => Array.isArray(obj);
