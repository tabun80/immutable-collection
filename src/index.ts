import { CollectionItemType } from './types';
import { deepFreeze } from './utils';

export type Collection<T extends CollectionItemType> = CollectionClass<T>;
class CollectionClass<T extends Object> {
  #items: readonly T[];

  constructor(items: T[]) {
    this.#items = [...items];
    deepFreeze(this);
  }

  *[Symbol.iterator]() {
    let index;
    for (index in this.#items) {
      yield this.#items[index];
    }
  }

  get length() {
    return this.#items.length;
  }

  toJson() {
    return [...this];
  }

  map<U>(callbackfn: (value: T, index: number, array: readonly T[]) => U) {
    const result = this.#items.map(callbackfn);

    return new Collection<U>(result);
  }

  filter(predicate: (value: T, index: number, array: readonly T[]) => unknown) {
    const result = this.#items.filter(predicate);

    return new Collection<T>(result as T[]);
  }
}

export const Collection: new <T extends CollectionItemType>(
  items: T[]
) => Collection<T> = CollectionClass as any;
