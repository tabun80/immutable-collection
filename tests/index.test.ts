import { Collection } from './../src';

type TestObject = {
  test: number;
};

class TestCollection extends Collection<TestObject> {}

describe('Collection class', () => {
  describe('property: length', () => {
    const mock = [{ test: 1 }, { test: 2 }, { test: 3 }];

    it('0', async () => {
      const data: TestObject[] = [];

      const collection = new TestCollection(data);
      expect(collection.length).toBe(0);
    });

    mock.forEach((_, index) => {
      it(String(index + 1), async () => {
        const collection = new TestCollection(mock.slice(0, index + 1));

        expect(collection.length).toBe(index + 1);
        expect(collection.toJson()).toMatchSnapshot();
      });
    });
  });

  describe('iterator', () => {
    const mock = [{ test: 1 }, { test: 2 }, { test: 3 }];

    it('return array', async () => {
      const collection = new TestCollection(mock);

      expect([...collection]).toMatchSnapshot();
    });
  });

  describe('method: toJson', () => {
    const mock = [{ test: 1 }, { test: 2 }, { test: 3 }];

    it('return array', async () => {
      const collection = new TestCollection(mock);

      expect(collection.toJson()).toMatchSnapshot();
    });
  });

  describe('method: map', () => {
    const mock = [{ test: 1 }, { test: 2 }, { test: 3 }];

    it('immutable', async () => {
      const collection = new TestCollection(mock);
      const result = collection.map(({ test }) => test + 1);

      expect(collection.toJson()).toMatchSnapshot();
      expect(result.toJson()).toMatchSnapshot();
    });
  });

  describe('method: filter', () => {
    const mock = [{ test: 1 }, { test: 2 }, { test: 3 }];

    it('immutable', async () => {
      const collection = new TestCollection(mock);
      const result = collection.filter(({ test }) => test === 1);

      expect(collection.toJson()).toMatchSnapshot();
      expect(result.toJson()).toMatchSnapshot();
    });
  });
});
