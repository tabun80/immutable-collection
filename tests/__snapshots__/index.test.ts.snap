// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection class iterator return array 1`] = `
Array [
  Object {
    "test": 1,
  },
  Object {
    "test": 2,
  },
  Object {
    "test": 3,
  },
]
`;

exports[`Collection class method: map immutable 1`] = `
Array [
  Object {
    "test": 1,
  },
  Object {
    "test": 2,
  },
  Object {
    "test": 3,
  },
]
`;

exports[`Collection class method: map immutable 2`] = `
Array [
  2,
  3,
  4,
]
`;

exports[`Collection class method: toJson return array 1`] = `
Array [
  Object {
    "test": 1,
  },
  Object {
    "test": 2,
  },
  Object {
    "test": 3,
  },
]
`;

exports[`Collection class property: length 1 1`] = `
Array [
  Object {
    "test": 1,
  },
]
`;

exports[`Collection class property: length 2 1`] = `
Array [
  Object {
    "test": 1,
  },
  Object {
    "test": 2,
  },
]
`;

exports[`Collection class property: length 3 1`] = `
Array [
  Object {
    "test": 1,
  },
  Object {
    "test": 2,
  },
  Object {
    "test": 3,
  },
]
`;
